File: java-automation/src/test/java/com/myecommerce/automation/driver/ports/HttpPort.java
package com.myecommerce.automation.driver.ports;

import java.util.List;

public interface HttpPort {

    <T> T getAs(String path, Class<T> type);
    <T> List<T> getListAs(String path, Class<T> elementType);
    <T> List<T> getListWithQueryAs(String path, String paramName, String paramValue, Class<T> elementType);
}

File: kotlin-automation/src/test/kotlin/com/myecommerce/automation/driver/ports/HttpPort.kt
package com.myecommerce.automation.driver.ports

interface HttpPort {
    fun <T : Any> getAs(path: String, type: Class<T>): T
    fun <T : Any> getListAs(path: String, elementType: Class<T>): List<T>
    fun <T : Any> getListWithQueryAs(path: String, paramName: String, paramValue: String, elementType: Class<T>): List<T>
}

inline fun <reified T : Any> HttpPort.getAs(path: String): T = getAs(path, T::class.java)
inline fun <reified T : Any> HttpPort.getListAs(path: String): List<T> = getListAs(path, T::class.java)
inline fun <reified T : Any> HttpPort.getListWithQueryAs(path: String, paramName: String, paramValue: String): List<T> =
    getListWithQueryAs(path, paramName, paramValue, T::class.java)
